### PR TITLE
#148 Add retrofit service for downloading apk files

### PR DESCRIPTION
**Pull request by tuancoltech**

### CODE REVIEW

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 1]
```diff
@@ -81,8 +86,13 @@ public DaoSession getDaoSession() {
 
     /**
      * Initializes Retrofit and makes it available to all activities.
+     *
+     * @param progressUpdateCallback If the request needs to update progress to user (in case of downloading or
+     *                               uploading
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Will the app be _uploading_ files?

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 2]
```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
...
                 HttpLoggingInterceptor.Level.NONE);
-        final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+
+        final ProgressResponseBody.ProgressListener progressListener = new ProgressResponseBody.ProgressListener() {
+            @Override
+            public void update(long bytesRead, long contentLength, boolean done) {
+//                Timber.i("bytesRead: " + bytesRead);
+//                Timber.i("contentLength: " + contentLength);
+//                Timber.i("done: " + done);
+//                Timber.i("%d%% done\n", (100 * bytesRead) / contentLength);
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Are these the lines that caused the OutOfMemoryException?

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 3]
```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
         HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
         logging.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY :
                 HttpLoggingInterceptor.Level.NONE);
-        final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+
+        final ProgressResponseBody.ProgressListener progressListener = new ProgressResponseBody.ProgressListener() {
+            @Override
+            public void update(long bytesRead, long contentLength, boolean done) {
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: How frequently is this method called? Every 1024 bytes?

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 1 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: In case we have to uploading a big file in the future, then we also can use this ProgressUpdateCallback for uploading progress.

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 4]
```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
...
+            }
+        };
+
+
+        final OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder()
                 .addInterceptor(logging)
                 .readTimeout(20, TimeUnit.SECONDS)
-                .retryOnConnectionFailure(true)
-                .build();
+                .retryOnConnectionFailure(true);
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: How many times will the retry happen? Once?

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 5]
```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
...
 
-        if (retrofit == null) {
+        /**
+         * In case of downloading/uploading big files and having necessity to update progress to UI, then add a
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Remove "uploading" from the Javadoc

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 2 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: No. But they are unnecessary IMHO, so just disabled them.

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java [hunk 6]
```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
...
+                    return originalResponse.newBuilder()
+                            .body(new ProgressResponseBody(originalResponse.body(), progressListener))
+                            .build();
+                }
+            });
+        }
+
+        OkHttpClient okHttpClient = okHttpClientBuilder.build();
+
+        if (retrofit == null || progressUpdateCallback != null) {
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Why is Retrofit initialized if `progressUpdateCallback != null`?

File: app/src/main/java/ai/elimu/appstore/service/ProgressResponseBody.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import okhttp3.MediaType;
+import okhttp3.ResponseBody;
+import okio.Buffer;
+import okio.BufferedSource;
+import okio.ForwardingSource;
+import okio.Okio;
+import okio.Source;
+
+/**
+ * HTTP response body which supports progress update when downloading/uploading big files
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/service/ProgressResponseBody.java]: Remove "uploading" from the text

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 3 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: It's not 1024 bytes but a random number, which is the number of bytes it receives from server.
Kindly refer to the class here
https://github.com/square/okhttp/blob/209c6c5589d7b0a0f10b4ce78ac6be0740a5cea8/samples/guide/src/main/java/okhttp3/recipes/Progress.java#L68

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 4 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Basically since we do not have many concurrent network requests, the maximum retry is 1.
In case we have many concurrent requests, then it depends on the thread pool from which requests are executed. Default number of retry in this case is 5.
Kindly refer to below comment

https://stackoverflow.com/a/47544627/3682198

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 6 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Each Retrofit instance is associated with a single OkHttpClient, which is also associated with only one single ProgressListener. 
When we have many concurrent downloads, each download should have a separate progress update callback. Therefore, it's necessary to initialize a new Retrofit connection for new download so that every download's progress is updated correctly.

When progressUpdateCallback is null, meaning that no progress update is required, then we can reuse existing Retrofit connection.

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 1 above)
Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Yes sure.

File: app/src/main/java/ai/elimu/appstore/service/ProgressResponseBody.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import okio.Buffer;
+import okio.BufferedSource;
+import okio.ForwardingSource;
+import okio.Okio;
+import okio.Source;
+
+/**
+ * HTTP response body which supports progress update when downloading big files
+ */
+public class ProgressResponseBody extends ResponseBody {
```
Comment (jo-elimuai) [inline app/src/main/java/ai/elimu/appstore/service/ProgressResponseBody.java]: Why is this file added if it's already included in the OkHttp library?: https://github.com/square/okhttp/blob/209c6c5589d7b0a0f10b4ce78ac6be0740a5cea8/samples/guide/src/main/java/okhttp3/recipes/Progress.java#L68

Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/service/ProgressResponseBody.java]: It's not included in OkHttp library.
The file above is within sample folder.

okhttp/samples/guide/src/main/java/okhttp3/recipes/Progress.java

File: app/src/main/java/ai/elimu/appstore/BaseApplication.java (hunk 6 above)
Comment (jo-elimuai) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Ok. Can you add this information to make the code easier to understand?

Comment (tuancoltech) [reply app/src/main/java/ai/elimu/appstore/BaseApplication.java]: Yep sure.
